### DISCUSSION TITLE
GraphQL

### DISCUSSION THREAD

**Discussion by Modcat** [feathersjs/feathers#2388] [Ideas]
I was wondering if GraphQL could be supported out of the box? Would add another good layer to it!

**Comment by daffl** [comment]
The new schemas and custom methods do allow for a pretty streamlined GraphQL transport integration, so definitely a valid suggestion. I might have more details on that soon but would like to take this opportunity to share some of my grievances around GraphQL:

- GraphQL is a protocol primarily pushed by Facebooks developer marketing budget. Its development process and marketing is essentially the opposite of how all the other internet protocol we use today evolved. I still feel it unfortunately has drowned out many other projects that are doing interesting and just as valuable work. Since it was designed for Facebook and other enterprise use-cases, I also wouldn't be surprised for it to eventually turn into a new version of [SOAP](https://simple.wikipedia.org/wiki/SOAP) in terms of complexity but with a custom DSL instead of XML.
- There are now open source companies backed by tens of millions of dollars in VC funding focussing specifically on this one protocol (especially to solving its drawbacks). This makes it not very motivating to work for free on what would essentially be an alternative. Unfortunately I'm not too optimistic that enough people and companies would support an initiative like this either financially or through contributions (there was a Feathers + GraphQL initiative which has been abandoned). At the least, I am planning to publish the Feathers GraphQL transport under a license other than MIT, probably AGPL which requires anybody who modifies the code to also publish it in the open (and not just run it as a paid closed service).
- GraphQL was designed in the mid 2010s which is why I'm still surprised it only includes real-time functionality as an afterthought in the form of arbitrary subscriptions. Feathers still seems to be the only library trying to provide overall architectural pattern for handling real-time data and who should be allowed to see them.

**Comment by Modcat** [comment]
So maybe my code define's a service like authenticate where a creates a single endpoint that has an API allowing me to to communicate like GraphQL but not totally, to multiple endpoints in one request, I know there is batch processing but I think it could be better
